ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+  // and a series of "maxFoo" limits to ensure the initialization flow doesn't get out of control.
+  private currentDelay: number;
+  private maxDelay: number;
+  private overallTimeout: number;
+  private maxCreateCount: number;
+  private maxDeleteCount: number;
+  private maxResumeCount: number;
+  private maxServerErrorCount: number;
+
+  // Properties to track progress, actions taken, and errors encountered.
+  private createCount = 0;
+  private deleteCount = 0;
+  private resumeCount = 0;
+  private serverErrorCount = 0;
+  private initializeStartTime?: number;
+  // The latest cluster retrieved from getCluster. If the last getCluster call returned a NOT_FOUND
+  // response, this will be null.
+  private currentCluster?: Cluster;
+
+  // Properties to control the initialization and promise resolution flow.
+  //
+  // Tracks whether the .run() method has been called yet on this class. Each instance is only
+  // allowed to be used for one initialization flow.
+  private isInitializing = false;
+  // The resolve and reject function from the promise returned from the call to .run(). We use a
+  // deferred-style approach in this class, which allows us to provide a Promise-based API on the
+  // .run() method, but to call the resolve() or reject() method from anywhere in this class.
+  private resolve: (cluster?: Cluster | PromiseLike<Cluster>) => void;
+  private reject: (error: Error) => void;
+
+  constructor(options: ClusterInitializerOptions) {
+    this.workspaceNamespace = options.workspaceNamespace;
+    this.onStatusUpdate = options.onStatusUpdate ? options.onStatusUpdate : () => {};
+    this.abortSignal = options.abortSignal;
+    this.currentDelay = options.initialPollingDelay != null ? options.initialPollingDelay : DEFAULT_INITIAL_POLLING_DELAY;
+    this.maxDelay = options.maxPollingDelay != null ? options.maxPollingDelay : DEFAULT_MAX_POLLING_DELAY;
+    this.overallTimeout = options.overallTimeout != null ? options.overallTimeout : DEFAULT_OVERALL_TIMEOUT;
+    this.maxCreateCount = options.maxCreateCount != null ? options.maxCreateCount : DEFAULT_MAX_CREATE_COUNT;
+    this.maxDeleteCount = options.maxDeleteCount != null ? options.maxDeleteCount : DEFAULT_MAX_DELETE_COUNT;
+    this.maxResumeCount = options.maxResumeCount != null ? options.maxResumeCount : DEFAULT_MAX_RESUME_COUNT;
+    this.maxServerErrorCount = options.maxServerErrorCount != null ? options.maxServerErrorCount : DEFAULT_MAX_SERVER_ERROR_COUNT;
+  }
+
+  private async getCluster(): Promise<Cluster> {
+    return await clusterApi().getCluster(this.workspaceNamespace, {signal: this.abortSignal});
+  }
+
+  private async createCluster(): Promise<Cluster> {
+    if (this.createCount >= this.maxCreateCount) {
+      throw new ExceededActionCountError('Reached max cluster create count', this.currentCluster);
+    }
+    const cluster = await clusterApi().createCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    this.createCount++;
```

Comment:
So if I'm reading this right, we increment before the `await` is complete? Is that what we want? Maybe it's really more of a `createAttemptCount`.

Reply:
🤦‍♂ right.